plugins/saveactions.c

```diff
@@ -258,49 +271,387 @@ static void backupcopy_document_save_cb(GObject *obj, GeanyDocument *doc, gpoint
+	g_free(msg);
+
+	switch (response)
+	{
+		case GTK_RESPONSE_YES:
+			old_file_path = g_strdup(file_path);
+
+			if (dialogs_show_save_as())
+			{
+				new_file_path = doc->file_name;
+
+				if (! g_str_equal(old_file_path, new_file_path))
+				{
+				 	/* remove temp file if it was saved as some other file */
+					g_remove(locale_file_path);
+				}
+			}
+			else
+			{
+				plugin_idle_add(geany_plugin, open_document_once_idle, g_strdup(locale_file_path));
+			}
+
+			g_free(old_file_path);
+			break;
+
+		case GTK_RESPONSE_NO:
+			g_remove(locale_file_path);
+
+			ui_set_statusbar(TRUE, _("Temp file %s was deleted"), short_filename);
+			break;
+
+		case GTK_RESPONSE_CANCEL:
+		default:
+			plugin_idle_add(geany_plugin, open_document_once_idle, g_strdup(locale_file_path));
+			break;
+	}
+}
+
+
+static void persistent_temp_files_document_close_cb(GObject *obj, GeanyDocument *doc, gpointer user_data)
+{
+	gchar *file_path, *locale_file_path, *short_filename;
+
+	file_path = doc->file_name;
+
+	if (enable_persistent_temp_files && file_path != NULL && ! geany_is_quitting())
+	{
+		if (is_temp_saved_file_doc(doc) && file_exists(doc->file_name))
+		{
+			short_filename = document_get_basename_for_display(doc, -1);
+			locale_file_path = utils_get_locale_from_utf8(file_path);
+
+			if (! document_is_empty(doc))
+			{
+				show_unsaved_dialog_for_persistent_temp_files_tab_closing(
+					doc,
+					file_path,
+					locale_file_path,
+					short_filename
+				);
+			}
+			else
+			{
+				g_remove(locale_file_path);
+
+				ui_set_statusbar(TRUE, _("Empty temp file %s was deleted"), short_filename);
+			}
+
+			g_free(short_filename);
+			g_free(locale_file_path);
+		}
+	}
+}
+
+
+static void persistent_temp_files_document_before_save_cb(GObject *obj, GeanyDocument *doc, gpointer user_data)
+{
+	if (enable_persistent_temp_files)
+	{
+		gchar *old_file_path, *new_file_path, *old_file_name;
+				GtkWidget *page_label = NULL;
+
+		get_current_tab_label(&page_label);
+
+		old_file_path = gtk_widget_get_tooltip_text(page_label);
+		new_file_path = DOC_FILENAME(doc);
+
+		if (old_file_path == NULL)
 		{
-			gchar *message = g_strdup_printf(
-				_("Instant Save filename could not be generated (%s)."), g_strerror(errno));
-			ui_set_statusbar(TRUE, "%s", message);
-			g_warning("%s", message);
-			g_free(message);
-			g_free(new_filename);
+			ui_set_statusbar(TRUE, _("plugin error: failed to delete initial temp file "
+				"('failed to get notebook tab label text')"));
+
 			return;
 		}
 
-		close(fd); /* close the returned file descriptor as we only need the filename */
+		old_file_name = g_path_get_basename(old_file_path);
+
+		if (is_temp_saved_file(old_file_name) && ! g_str_equal(old_file_path, new_file_path))
+		{
+			/* we have to store old/new filename pair in a global hashtable to be able to somehow
+			pass it to document-save callback that is called directly after this one */
+			g_hash_table_insert(
```

Comment:
we have to use this hack with hashtable to remember name of original 'temp' file, during performing 'save as' which assigns a new 'final' name to file.

Ideally we would like to do all these actions in a more proper manner but this would require core code modifications

Reply:
done